File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {    
    int[][] directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        
    private void dfs(int r, int c, boolean[][] ocean, int[][] heights){
            ocean[r][c] = true;
            for(int[] dir: directions){
                int nr = r + dir[0];
                int nc = c + dir[1];
                if((nr >= 0 && nr < heights.length)
                    && (nc >= 0 && nc < heights[0].length)
                    && !ocean[nr][nc] && heights[nr][nc] >= heights[r][c]){
                        dfs(nr, nc, ocean, heights);
                    }
            }
    }

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> result = new ArrayList<>();
        int rows = heights.length;
        int cols = heights[0].length;
        boolean[][] visitedP = new boolean[rows][cols];
        boolean[][] visitedA = new boolean[rows][cols];

        for(int r = 0; r < rows; r++){
            dfs(r, 0, visitedP, heights);
            dfs(r, cols - 1, visitedA, heights);
        }

        for(int c = 0; c < cols; c++){
            dfs(0, c, visitedP, heights);
            dfs(rows - 1, c, visitedA, heights);
        }
        for(int r = 0; r < rows; r++)
            for(int c = 0; c < cols; c++)
                if(visitedP[r][c] && visitedA[r][c])
                    result.add(Arrays.asList(r, c));
        return result;
    }
}

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        ROWS, COLS = len(heights), len(heights[0])
        #two sets containing coordinates of cells which can go to pacific and atlantic oceans
        pac, atl = set(), set()
        directions = [[0, 1], [1, 0], [-1, 0], [0, -1]]
        def dfs(r, c, visited, prevHeight):
            if r not in range(ROWS) or c not in range(COLS) or (r, c) in visited or heights[r][c] < prevHeight:
                return
            visited.add((r, c))
            for dr, dc in directions:
                nr, nc = dr + r, dc + c
                dfs(nr, nc, visited, heights[r][c])
        for r in range(ROWS):
            dfs(r, 0, pac, heights[r][0])
            dfs(r, COLS - 1, atl, heights[r][COLS - 1])
        for c in range(COLS):
            dfs(0, c, pac, heights[0][c])
            dfs(ROWS - 1, c, atl, heights[ROWS - 1][c])
        res = []
        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) in pac and (r, c) in atl:
                    res.append([r, c])
        return res